pennylane/devices/default_qubit.py

```diff
@@ -438,6 +448,38 @@ def _apply_phase(self, state, axes, parameters, inverse=False):
         phase = self._conj(parameters) if inverse else parameters
         return self._stack([state[sl_0], phase * state[sl_1]], axis=axes[0])
 
+    def expval(self, observable, shot_range=None, bin_size=None):
+        """Returns the expectation value of a Hamiltonian observable. When the observable is a
+         ``SparseHamiltonian`` object, the expectation value is computed directly for the full
+         Hamiltonian, which leads to faster execution.
+
+        Args:
+            observable (~.Observable): a PennyLane observable
+            shot_range (tuple[int]): 2-tuple of integers specifying the range of samples
+                to use. If not specified, all samples are used.
+            bin_size (int): Divides the shot range into bins of size ``bin_size``, and
+                returns the measurement statistic separately over each bin. If not
+                provided, the entire shot range is treated as a single bin.
+
+        Returns:
+            float: returns the expectation value of the observable
+        """
+        if observable.name == "SparseHamiltonian" and self.shots is not None:
+            raise DeviceError("SparseHamiltonian must be used with shots=None")
+
+        if observable.name == "SparseHamiltonian" and self.shots is None:
```

Comment:
Actually, you could combine it:

```python
if observable.name == "SparseHamiltonian":
    if self.shots is not None:
        raise DeviceError("SparseHamiltonian must be used with shots=None")

    ev = coo_matrix.dot(
```

This may be _very slightly_ more efficient, just be reducing the number of if statements that are computed on each evaluation.